src/core/Type.ml

```diff
@@ -0,0 +1,211 @@
+(* Yoann Padioleau
+ *
+ * Copyright (C) 2022, 2023 r2c
+ *
+ *)
+open Common
+module G = AST_generic
+
+let logger = Logging.get_logger [ __MODULE__ ]
+
+(*****************************************************************************)
+(* Prelude *)
+(*****************************************************************************)
+(* Types for type inference.
+ *
+ * Why not simply reuse AST_generic.type_ ? Because here we want to make
+ * sure type names are fully resolved! AST_generic is also meant to represent
+ * types that people could actually write as part of programs. This is a more
+ * abstract representation of types, divorced from the notion of a syntax tree.
+ * For example, we don't include tokens here. We may introduce types here that
+ * wouldn't appear as written types in a program. We might even introduce type
+ * variables here, which would not be appropriate to include in AST_generic.
+ *
+ * In the future, we may also want to apply more complex normalizations to
+ * simplify typing/naming.
+ *
+ * TODO: polymorphic types? type parameters are resolved names?
+ *
+ * These types are polymorphic to support the Pro Engine, which represents
+ * resolved names differently.
+ *)
+
+(*****************************************************************************)
+(* Types *)
+(*****************************************************************************)
+type todo_kind = string option [@@deriving show]
+
+(* Fully qualified name *)
```

Comment:
I'm not a fan of repeating potentially-large variant types in an `mli`. Especially because in this case I anticipate generating visitors for it at some point, which aren't compatible with an `mli`.

Reply:
Oh wait, I already wrote a comment that explains the polymorphism.